File: .tmp_py4013/batch_17_dump/3878_count_good_subarrays.js
// LeetCode 3878 - Count Good Subarrays
// https://leetcode.com/problems/count-good-subarrays/

var countGoodSubarrays = function(nums) {
    const n = nums.length;
    const l = new Array(n).fill(-1);
    const stk = [];
    for (let i = 0; i < n; i++) {
        const x = nums[i];
        while (stk.length > 0 && nums[stk[stk.length - 1]] < x && (nums[stk[stk.length - 1]] | x) === x) {
            stk.pop();
        }
        if (stk.length > 0) l[i] = stk[stk.length - 1];
        stk.push(i);
    }
    const r = new Array(n).fill(n);
    stk.length = 0;
    for (let i = n - 1; i >= 0; i--) {
        while (stk.length > 0 && (nums[stk[stk.length - 1]] | nums[i]) === nums[i]) {
            stk.pop();
        }
        if (stk.length > 0) r[i] = stk[stk.length - 1];
        stk.push(i);
    }
    let ans = 0;
    for (let i = 0; i < n; i++) {
        ans += (i - l[i]) * (r[i] - i);
    }
    return ans;
};

File: .tmp_rs4013/rust_check21/3878_count_good_subarrays.rs
struct Solution;
// LeetCode 3878 - Count Good Subarrays
// https://leetcode.com/problems/count-good-subarrays/

impl Solution {
    pub fn count_good_subarrays(nums: Vec<i32>) -> i64 {
        let n = nums.len();
        let mut l = vec![-1i32; n];
        let mut stk = Vec::new();
        for i in 0..n {
            let x = nums[i];
            while !stk.is_empty() {
                let last = *stk.last().unwrap();
                if nums[last] < x && (nums[last] | x) == x {
                    stk.pop();
                } else {
                    break;
                }
            }
            if !stk.is_empty() {
                l[i] = *stk.last().unwrap() as i32;
            }
            stk.push(i);
        }
        let mut r = vec![n as i32; n];
        stk.clear();
        for i in (0..n).rev() {
            while !stk.is_empty() {
                let last = *stk.last().unwrap();
                if (nums[last] | nums[i]) == nums[i] {
                    stk.pop();
                } else {
                    break;
                }
            }
            if !stk.is_empty() {
                r[i] = *stk.last().unwrap() as i32;
            }
            stk.push(i);
        }
        let mut ans = 0i64;
        for i in 0..n {
            ans += (i as i32 - l[i]) as i64 * (r[i] - i as i32) as i64;
        }
        ans
    }
}
